### PR TITLE
Use elasticsearch to get annotation_user_count

### PR DESCRIPTION
**Pull request by hmstepanek**
Previously the count was obtained from a db query. The problem with
this was two fold:
 1. Shared was being queried on even though it wasn't indexed
 2. When searching for users such as scibot, where the user has a
    large portion of the annotations in the table, the db was optimizing
    to search the table directly rather than use the index due to the
    overhead of indexing lookup being larger than a full table scan.

Elasticsearch performs much better for aggregation queries with 
complex where clauses and it also has built in caching. Although
looking at the scibot user on activity pages doesn't happen that 
often, when it does, it takes up to 7.5 s of query time to perform 
this lookup. ES ends up taking ~.02 s. 

Rename annotation_user_counts to annotation_user_count and simply
return the total as opposed to the previous implementation that 
returned 3 different values. 

Add a new method total_annotation_user_counts that counts all the user's
annotations regardless of permissions, public/private, and includes replies.

Cache the parsed query params from query.extract so that the parser
doesn't have to be run twice in the group and user controllers.

If able, re-use the total user counts from the search to avoid running
the same query twice. 

This is the final fix for https://github.com/hypothesis/product-backlog/issues/812.

### CODE REVIEW

File: h/services/annotation_stats.py [hunk 1]
```diff
@@ -16,36 +14,27 @@ class AnnotationStatsService(object):
...
-
-        annotations = self.request.db.query(Annotation). \
-            filter_by(userid=userid, deleted=False). \
-            options(sa.orm.load_only('groupid', 'shared')).subquery()
-        grouping = sa.case([
-            (sa.not_(annotations.c.shared), 'private'),
-            (annotations.c.groupid == '__world__', 'public'),
-        ], else_='group')
-
-        result = dict(self.request.db.query(grouping, sa.func.count(annotations.c.id)).group_by(grouping).all())
```
Comment (robertknight) [inline h/services/annotation_stats.py]: Do you know what the generated SQL looked like for this query. The initial `annotations` query doesn't look inefficient (`SELECT groupid, shared FROM annotation WHERE userid := userid, deleted = false`), but I can't immediately see how the rest of the SQLAlchemy query translates to an actual DB query.

Comment (hmstepanek) [reply h/services/annotation_stats.py]: SELECT CASE WHEN NOT anon_2.shared THEN %(param_1)s WHEN (anon_2.groupid = %(groupid_1)s) THEN %(param_2)s ELSE %(param_3)s END AS anon_1, count(anon_2.id) AS count_1 FROM (SELECT annotation.id AS id, annotation.groupid AS groupid, annotation.shared AS shared FROM annotation WHERE annotation.deleted = ? AND annotation.userid = %(userid_1)s) AS anon_2 GROUP BY CASE WHEN NOT anon_2.shared THEN %(param_1)s WHEN (anon_2.groupid = %(groupid_1)s) THEN %(param_2)s ELSE %(param_3)s END

File: h/views/admin/users.py [hunk 2]
```diff
@@ -37,8 +37,7 @@ def users_index(request):
 
     if user is not None:
         svc = request.find_service(name='annotation_stats')
-        counts = svc.user_annotation_counts(user.userid)
-        user_meta['annotations_count'] = counts['total']
+        user_meta['annotations_count'] = svc.user_annotation_count(user.userid)
```
[the remark this answers is not in the document]
Comment (robertknight) [reply h/views/admin/users.py]: The query I mentioned above though can be satisfied using a simple index-only scan:

```
postgres=# explain select count(*) from annotation where userid = 'foo';
                                            QUERY PLAN                                             
---------------------------------------------------------------------------------------------------
 Aggregate  (cost=8.16..8.17 rows=1 width=0)
   ->  Index Only Scan using ix__annotation_userid on annotation  (cost=0.14..8.16 rows=1 width=0)
         Index Cond: (userid = 'foo'::text)
(3 rows)

postgres=# 
```

I anticipate this would be even faster than querying Elasticsearch.

[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/views/admin/users.py]: In the case where there aren't a lot of annotations for a user, the query is fast but that's not the case I'm trying to fix here, I'm trying to fix the case where there are a tone of annotations and the query is extremely slow.
image
image

Here's an example query from scibot:
image
image

I think it's also worth considering we won't be running this query at all in most cases if we make this change thus it will always be faster. (I haven't seen a slow query that has other search criteria in it so generally I think it's fair to say this query will almost always be skipped.)

**Comment by hmstepanek** [comment]
@robertknight I've added the new method for admin pages so PTAL when you get a chance.

File: h/services/annotation_stats.py [hunk 3]
```diff
@@ -16,36 +14,45 @@ class AnnotationStatsService(object):
...
 
-        return result
+        search_result = search.run(params)
+        return search_result.total
 
     def group_annotation_count(self, pubid):
         """
         Return the count of searchable top level annotations for this group.
         """
+        params = MultiDict({'limit': 0, 'group': pubid})
```
Comment (robertknight) [inline h/services/annotation_stats.py]: It looks like we're always passing `"limit": 0` when calling `_search`, so this parameter could be "pushed down" into that method.

File: h/views/activity.py [hunk 4]
```diff
@@ -35,10 +35,13 @@ class SearchController(object):
 
     def __init__(self, request):
         self.request = request
+        # Cache a copy of the extracted query params for the child controllers to use if needed.
+        self.parsed_query_params = query.extract(self.request)
```
Comment (robertknight) [inline h/views/activity.py]: Why is this necessary?

Comment (hmstepanek) [reply h/views/activity.py]: And this is necessary because search eats the params.
